payments/providers/bambora_payform.py

```diff
@@ -0,0 +1,313 @@
+            return self.ui_redirect_failure()
+
+        try:
+            order = Order.objects.get(order_number=request.GET["ORDER_NUMBER"])
+        except Order.DoesNotExist:
+            logger.warning("Order does not exist.")
+            return self.ui_redirect_failure()
+
+        return_code = request.GET["RETURN_CODE"]
+        if return_code == "0":
+            logger.debug("Payment completed successfully.")
+            try:
+                order.set_status(
+                    OrderStatus.PAID,
+                    "Code 0 (payment succeeded) in Bambora Payform success request.",
+                )
+                return self.ui_redirect_success(order)
+            except OrderStatusTransitionError as oste:
+                logger.warning(oste)
+                order.create_log_entry(
+                    "Code 0 (payment succeeded) in Bambora Payform success request."
+                )
+                return self.ui_redirect_failure(order)
+        elif return_code == "1":
+            logger.debug("Payment failed.")
+            try:
+                order.set_status(
+                    OrderStatus.REJECTED,
+                    "Code 1 (payment rejected) in Bambora Payform success request.",
+                )
+                return self.ui_redirect_failure(order)
+            except OrderStatusTransitionError as oste:
+                logger.warning(oste)
+                order.create_log_entry(
+                    "Code 1 (payment rejected) in Bambora Payform success request."
+                )
+                return self.ui_redirect_failure(order)
+        elif return_code == "4":
+            logger.debug("Transaction status could not be updated.")
+            order.create_log_entry(
+                "Code 4: Transaction status could not be updated. Use the merchant UI to resolve."
+            )
+            return self.ui_redirect_failure(order)
+        elif return_code == "10":
+            logger.debug("Maintenance break.")
+            order.create_log_entry("Code 10: Bambora Payform maintenance break")
+            return self.ui_redirect_failure(order)
+        else:
+            logger.warning('Incorrect RETURN_CODE "{}".'.format(return_code))
+            order.create_log_entry(
+                'Bambora Payform incorrect return code "{}".'.format(return_code)
+            )
+            return self.ui_redirect_failure(order)
+
+    def handle_notify_request(self):
+        """Handle the asynchronous part of payform response
+
+        Arrives some time after user has completed the payment flow or stopped it abruptly.
+        Skips changing order status if it has been previously set. Although, according to
+        Bambora's documentation, there are some cases where payment status might change
+        from failed to successful, the reservation has probably been soft-cleaned up by then.
+        TODO Maybe try recreating the reservation if the time slot is still available
```

Comment:
I think we don't soft-clean our orders. I guess, if first API call to Bambora said that transaction did not go through, but this notify call says so, we can just mark the order as paid. 🤔

Reply:
... which is what this function does. so, just delete these couple lines from this docstring?

Reply:
Alright 🤷

Reply:
So, should we allow transition from cancelled to paid? In case the notify would mark that transition.

Reply:
I was thinking about that, but did not fully get the flow. Can it happen so, that first Bambora says "Failed" and we mark it as "Rejected" or "Canceled" but then a couple seconds later it sends us a notification that "Hey, actually it went ok"?
If that is the case, we should probably allow going from "Canceled" to "Paid"

Reply:
According to Bambora docs, it would be rare but possible:

> _In some rare cases it's possible that the status of the transaction changes from failed to successful later. In such cases the notify request is done again._

Reply:
ok, then we should allow going from `REJECTED` to `PAID` for the sake of this notify call 👍

Reply:
There would also be the case when a payment was rejected because something went wrong with the payment itself. For that I think we should allow moving from `REJECTED -> WAITING` so the user can try and pay again, just not sure how we could "enable" this somehow for the user.

Reply:
hmm, yeah, if that can happen, going `REJECTED -> WAITING` should be possible